### PR TITLE
[Freyja] - Update to version v1.5.1 and make it compatible with ILMN PE, ILMN SE and ONT

### LINKED ISSUE #546: Add additional Freyja args for ONT compatibility
:cool:

### :pushpin: Explain the Request

This feature will require a little additional exploration to determine how useful it would be.

Freyja has some additional input arguments that our workflow does not currently provide the ability to modify. Specifically, the --adapt input which "should suppress low level false positives" if set to the expected error rate.

While we're at it, could we update to the most recent version of Freyja?

### PR DESCRIPTION
**Pull request by cimendes**
This PR closes #546

🗑️ This dev branch should <NOT> be deleted after merging to main.

## :brain: Aim, Context and Functionality

This PR reworks the `Freyja_FASTQ` workflow to be technology-agnostic, allowing analysis with sequencing data from Illumina (paired-end and single-end) and ONT instruments (the first of it's kind!!!). 

Additionally, Freyja was updated to the latest version, `v1.5.1`:
- The barcodes file now is expected to be in the `.feather` format (see [here](https://github.com/andersen-lab/Freyja/releases/tag/v1.5.1))
- Additional optional inputs have been exposed for `freyja demix`

## :hammer_and_wrench:  Impacted Workflows/Tasks & Changes Being Made
- This will affect the behavior of the workflow(s) even if users don’t change any workflow inputs relative to the last version : No if update if turned off and the same type of data is used.

- Running this workflow on different occasions could result in different results, e.g. due to use of a live database, "latest" docker image, or stochastic data processing : Yes only if update is true 

For **Illumina**, the following logic is implemented:
- If a `read2` is provided, the data is QCed with the `workflows/utilities/wf_read_QC_trim_pe.wdl` sub-workflow otherwise, it uses the `workflows/utilities/wf_read_QC_trim_se.wdl` sub-workflow
- The clean read (or reads) are aligned to the provided reference genome with the `bwa` task (already compatible to optionally receive the `read2` file)

For **ONT**, activated by passing the `ont` boolean, the following logic is implemented:
- If `ont`, the length of the provided reference is calculated with `seqkit` via the `get_fasta_genome_size` task
- The calculated reference length and raw read is passed to the `nanoplot_task.nanoplot` task for quality assessment
- The data is QCed with the `workflows/utilities/wf_read_QC_trim_ont.wdl` sub-workflow
- The calculated reference length and clean read is passed to the `nanoplot_task.nanoplot` task for quality assessment
- The clean read is aligned to the provided reference genome with the `minimap2` task
- The resulting sam file is converted to bam and sorted with the `task_parse_mapping.sam_to_sorted_bam` task

The rest of the steps are followed regardless of the type of input data (unchanged from how it was functioning before this PR):
- primers are trimmed from the bam file with the `trim_primers.primer_trim` task
- all the Freyja magic happens in the `freyja_task.freyja_one_sample` task

#### 🔄 Data Processing 

Docker/software or software versions changed: `freyja:1.4.8` -> `freyja:1.5.1-07_02_2024-01-27-2024-07-22`

Databases or database versions changed: None

Data processing/commands changed: None

File processing changed: See above for the new tracks added for Illumina SE and ONT data

Compute resources changed: None

#### ➡️ Inputs 

New input:
- `ont` (default false)
- `freyja.adapt` (optional)

Downgraded to optional:
- `read2`

#### ⬅️ Outputs 

- `fastp_html_report`
- `fastp_version`
- `fastq_scan_num_reads_clean1`
- `fastq_scan_num_reads_clean2`
- `fastq_scan_num_reads_clean_pairs`
- `fastq_scan_num_reads_raw1`
- `fastq_scan_num_reads_raw2`
- `fastq_scan_num_reads_raw_pairs`
- `fastqc_docker`
- `freyja_lineage_metadata_file` <- New and improved!
- `freyja_usher_barcode_file` <- New and improved!
- `minimap2_docker`
- `minimap2_version`
- `nanoplot_html_clean`
- `nanoplot_html_raw`
- `nanoplot_num_reads_clean1`
- `nanoplot_num_reads_raw1`
- `nanoplot_r1_est_coverage_clean`
- `nanoplot_r1_est_coverage_raw`
- `nanoplot_r1_mean_q_clean`
- `nanoplot_r1_mean_q_raw`
- `nanoplot_r1_mean_readlength_clean`
- `nanoplot_r1_mean_readlength_raw`
- `nanoplot_r1_median_q_clean`
- `nanoplot_r1_median_q_raw`
- `nanoplot_r1_median_readlength_clean`
- `nanoplot_r1_median_readlength_raw`
- `nanoplot_r1_n50_clean`
- `nanoplot_r1_n50_raw`
- `nanoplot_r1_stdev_readlength_clean`
- `nanoplot_r1_stdev_readlength_raw`
- `nanoplot_tsv_clean`
- `nanoplot_tsv_raw`
- `nanoq_version`
- `trimmomatic_docker`

#### Commandline Testing with MiniWDL or Cromwell (optional)

- **freyja_update_refs task**
image

- **freyja_fastq workflow**
  - Illumina PE
image
image
  - ONT
image
image

#### Terra Testing

- **Freyja_FASTQ** Illumina Paired-End with update DB set to `true`: https://app.terra.bio/#workspaces/theiagen-validations/Theiagen_Mendes_Sandbox/job_history/04efacc2-6a99-49dd-828e-75b4f25b3a8c

- **Freyja_FASTQ** ONT with update DB set to `true` and ont set to `true`: https://app.terra.bio/#workspaces/theiagen-validations/Theiagen_Mendes_Sandbox/job_history/8b243c25-d9fe-4689-8b71-8ea815092e90 (many samples failed due to bad data (no reads left after QC)

- **Freyja_Update**: https://app.terra.bio/#workspaces/theiagen-validations/Theiagen_Mendes_Sandbox/job_history/e1eecc93-29b1-4ae5-867e-88315b593754

- **Freyja_Plot**: https://app.terra.bio/#workspaces/theiagen-validations/Theiagen_Mendes_Sandbox/job_history/bf7968b2-54be-4c3b-a15d-4c2a24f51761

- **Freyja_Dashboard**: Didn't test due to missing data. 😢 

## :microscope: Final Developer Checklist

- [x] The workflow/task has been tested locally and results, including file contents, are as anticipated
- [x] The workflow/task has been tested on Terra and results, including file contents, are as anticipated
- [x] The CI/CD has been adjusted and tests are passing (to be completed by Theiagen developer)
- [x] Code changes follow the [style guide](https://theiagen.notion.site/Style-Guide-WDL-Workflow-Development-bb456f34322d4f4db699d4029050481c)

## 🎯 Reviewer Checklist 

- [ ] All impacted workflows/tasks have been tested on Terra with a different dataset than used for development
- [ ] All reviewer-suggested scenarios have been tested and any additional
- [ ] All changed results have been confirmed to be accurate
- [ ] All workflows/tasks impacted by change/s have been tested using a standard validation dataset to ensure no unintended change of functionality
- [x] All code adheres to the style guide
- [x] MD5 sums have been updated
- [x] The PR author has addressed all comments

## 🗂️ Associated Documentation (to be completed by Theiagen developer)

- [x] Relevant documentation on the Public Health Resources "PHB Main" has been updated
- [ ] Workflow diagrams have been updated to reflect changes

### CODE REVIEW

**Comment by cimendes** [comment]
Question: Add nanoplot of not?

**Comment by cimendes** [comment]
Update Freyja in all workfows!

**Comment by michellescribner** [comment]
@cimendes documenting that we should update the freyja_barcode_version and freyja_metadata_version columns to be more specific about the file that was used, not just the data the workflow was run

**Comment by cimendes** [comment]
> @cimendes documenting that we should update the freyja_barcode_version and freyja_metadata_version columns to be more specific about the file that was used, not just the data the workflow was run

Now exposing both barcode and lineage files as an output for freyja_fastq!

**Comment by cimendes** [comment]
Possible improvement:
Alignment method is not 100% accurate for Illumina data and is just wrong for ONT data
image

**Comment by cimendes** [comment]
@frankambrosio3 checked with the dev and a barcode.csv should still be accepted, but i haven't tested

**Comment by frankambrosio3** [comment]
Tested 18 ONT samples and 20 Illumina samples. All fairly recent wastewater metagenomics amplicon artic v5.3.2 samples:
https://app.terra.bio/#workspaces/theiagen-demos/Ambrosio_Demo/job_history

ONT Run: https://app.terra.bio/#workspaces/theiagen-demos/Ambrosio_Demo/job_history/06df0bc8-874d-45a7-9273-2dc43cdb361c
In ONT run 4/18 samples failed:
SRR28240322 
SRR28240325 
SRR28240326 
SRR28240328 

File sizes not out of range compared to successful samples.

All failed at freyja task. All had the following error message:
image

Job manager log file for each sample:
SRR28240322 
https://job-manager.dsde-prod.broadinstitute.org/jobs/f5a59c7f-0568-48dd-bd29-c9796710d58c
image


SRR28240325 
https://job-manager.dsde-prod.broadinstitute.org/jobs/3af7ed97-0c55-4259-b100-a5ccd997e1c6
image


SRR28240326 
https://job-manager.dsde-prod.broadinstitute.org/jobs/bda570aa-b19b-4954-9251-f579efee0e85
image


SRR28240328 
https://job-manager.dsde-prod.broadinstitute.org/jobs/4e1f874c-33a7-4ff4-a2cb-0d4ac2c531d9
image


Looks like there was an issue with the freyja_demixed file generation for these four.

Execution directory file for each:
SRR28240322 
https://console.cloud.google.com/storage/browser/fc-d6683a3d-664d-44be-8d04-d66d139e8db5/submissions/06df0bc8-874d-45a7-9273-2dc43cdb361c/freyja_fastq/f5a59c7f-0568-48dd-bd29-c9796710d58c/call-freyja/attempt-4;tab=objects?authuser=0&prefix=&forceOnObjectsSortingFiltering=false

SRR28240325 
https://console.cloud.google.com/storage/browser/fc-d6683a3d-664d-44be-8d04-d66d139e8db5/submissions/06df0bc8-874d-45a7-9273-2dc43cdb361c/freyja_fastq/3af7ed97-0c55-4259-b100-a5ccd997e1c6/call-freyja/attempt-4;tab=objects?authuser=0&prefix=&forceOnObjectsSortingFiltering=false

SRR28240326 
https://console.cloud.google.com/storage/browser/fc-d6683a3d-664d-44be-8d04-d66d139e8db5/submissions/06df0bc8-874d-45a7-9273-2dc43cdb361c/freyja_fastq/bda570aa-b19b-4954-9251-f579efee0e85/call-freyja/attempt-4?pageState=(%22StorageObjectListTable%22:(%22f%22:%22%255B%255D%22))&authuser=0

SRR28240328 
https://console.cloud.google.com/storage/browser/fc-d6683a3d-664d-44be-8d04-d66d139e8db5/submissions/06df0bc8-874d-45a7-9273-2dc43cdb361c/freyja_fastq/4e1f874c-33a7-4ff4-a2cb-0d4ac2c531d9/call-freyja/attempt-4?authuser=0&pageState=(%22StorageObjectListTable%22:(%22f%22:%22%255B%255D%22))



Illumina Run: https://app.terra.bio/#workspaces/theiagen-demos/Ambrosio_Demo/job_history/7d9b1369-6c2f-44fd-bb34-4c3b482f32ad
At the time of writing this, 17 out of 20 successfully completed, looking like the other 3 will fail because 2 are on 3rd attempts at freyja task (SRR26423096 and SRR26423100) and the other one is still on primer trimming (SRR26266579).
Illumina fails:
SRR26266579
SRR26423096 
SRR26423100

Illumina fails job managers and logs:
SRR26266579
https://job-manager.dsde-prod.broadinstitute.org/jobs/31ac1134-cdf4-46e2-bba1-a3a94a4085c9
image


SRR26423096 
https://job-manager.dsde-prod.broadinstitute.org/jobs/c8f12635-4e49-43e9-855f-23417adb6676
image
^This again!

SRR26423100
https://job-manager.dsde-prod.broadinstitute.org/jobs/dab79239-3189-459e-bc17-e0ba0f81da2c
image

**Comment by cimendes** [comment]
Regarding your Illumina samples, the key error feels like it might be solved by forcing the database to update. I've launched a test locally with sample **SRR26266579** and it was successful:
image
image

@frankambrosio3 I'm going to relaunch just the failed samples in the Illumina dataset with the update database option set to true!

**Comment by cimendes** [comment]
Illumina test with update database set to true: https://app.terra.bio/#workspaces/theiagen-demos/Ambrosio_Demo/job_history/db7b3789-2111-4343-928b-11aa85cd3262

**Review by sage-wright — COMMENTED** [review]
code changes look good. will approve & merge upon signal from @cimendes

File: workflows/freyja/wf_freyja_fastq.wdl [hunk 1]
```diff
@@ -1,41 +1,98 @@
...
+        read1 = select_first([read_QC_trim_pe.read1_clean, read_QC_trim_se.read1_clean]),
+        read2 = select_first([read_QC_trim_pe.read2_clean])
+    }
   }
   call trim_primers.primer_trim {
     input:
       samplename = samplename,
       primer_bed = primer_bed,
-      bamfile = bwa.sorted_bam
+      bamfile = select_first([sam_to_sorted_bam.bam,bwa.sorted_bam])
```
[the remark this answers is not in the document]
Comment (cimendes) [reply workflows/freyja/wf_freyja_fastq.wdl]: done!

**Review by sage-wright — APPROVED** [review]

**Review by sage-wright — APPROVED** [review]
Good sleuthing!!!

**Comment by cimendes** [comment]
**Regarding ONT samples:** 
- sometimes the read can map outside the reference region, causing Freyja to fail. This can be addressed by changing the mapping params with minimap2

**Pull request merged**